llvm/lib/CAS/ActionCache.cpp

```diff
@@ -20,3 +21,5 @@ CacheKey::CacheKey(const ObjectProxy &Proxy)
     : CacheKey(Proxy.getCAS(), Proxy.getRef()) {}
 CacheKey::CacheKey(const ObjectStore &CAS, const ObjectRef &Ref)
     : Key(toStringRef(CAS.getID(Ref).getHash())) {}
+CacheKey::CacheKey(ArrayRef<uint8_t> Data)
+    : Key(toStringRef(BLAKE3::hash(Data))) {}
```

Comment:
This looks very wrong. Why do we need a CacheKey for any char array? And you can't hash the data before converting to CacheKey, the hashing needs to happen inside the implementation so you won't have two different cache key colliding.

Reply:
> Why do we need a CacheKey for any char array?

This is to be able to implement the remote key-value API using a `cas::ActionCache`, without having to store each individual key in a `cas::ObjectStore`, just to be able to get a `CASID` to use for key.

> And you can't hash the data before converting to CacheKey, the hashing needs to happen inside the implementation so you won't have two different cache key colliding.

It's not clear to me what you mean, could you clarify? With "cache key colliding" do you mean getting the same hash for different set of data? That would mean that we broke the BLAKE3 algorithm and we have bigger problems.

Reply:
The data stored in CacheKey is the raw input key. Let's say you somehow mixed CASID and ArrayRef<uint8_t> data and try to store in ActionCache as CacheKey.

CASID is a BLAKE3 hashed value and it is probably computed from another ArrayRef<> Buffer. Now if you store Buffer into the same action cache, Buffer will gets BLAKE3 hashed and it shared the same entry as the CASID computed from buffer.

Reply:
To make sure I understand, the concern is that:

* I have some ArrayRef data `data1`, that I store in an `ObjectStore` and get a `CASID` that I use as the `CacheKey` for the `ActionCache`; let's call it `key1`.
* I also have some ArrayRef data `data2` that I pass to `CacheKey` directly and that results in `key2`.
* Due to using BLAKE3 for both buffers, if `data1` == `data2` then `key1` == `key2` and we have an undesirable collision of `CacheKey`s.

Did I understand the concern correctly? If yes, then what do you recommend for change?
Should I include some additional static string, like `"__cachekey_from_arrayref__"`, in the hash value of the `ArrayRef<uint8_t>` buffer to avoid this `CacheKey` conflict?

Reply:
Correct. We need to avoid creating cache key that gets hashed different number of times. It might be fine to do that as implementation details for a test ActionCache but I would not add that to the general API for ActionCache.

Here are some ways to avoid the problem.
* Hash CASID and ArrayRef<uint8_t> one more time before inserting into ActionCache. The original ActionCache interface has this feature so we can probably bring it back if we want to use things other than CASID as key. It would be good to show that we have a real world use case for this, other than the tests.
* Insert ArrayRef into an ObjectStore to get an CASID before insert into ActionCache. You can use a InMemoryCAS, or even implement a nullCAS, which just produce the CASID and threw away the storage.
* For testing, you can just pad your string into desired length and pretend it is a CAS key, or just create a new CASContext for testing (let's say it requires key to be 4 bytes and you can just `CASID::create` them out of thin air) and create a world of ObjectStore/ActionCache out of this world.

I think all of them are acceptable but if you do 1, we need to have a discussion and implement that in a separate PR.

Reply:
> Due to using BLAKE3 for both buffers, if data1 == data2 then key1 == key2 and we have an undesirable collision of CacheKeys.

Hmm, actually the keys in this case are going to be different, because `BuiltinObjectHasher` also hashes the number of references (in this case `0`), so just hashing the data will result in a different key. Does this change whether there is a concern or not?

> Hash CASID and ArrayRef<uint8_t> one more time before inserting into ActionCache

It's not clear to me why we need to hash twice, wouldn't it be enough to have the `CASContext` contributing a static string for hashing (e.g. "llvmcas"), so that we avoid conflicts between different `CASContext`s? We can treat a standalone buffer as a separate hashing `CASContext` (with a different static string). In any case, we can discuss more in another PR.

> It would be good to show that we have a real world use case for this, other than the tests.

I think this PR does demonstrate a real world use case; I want to implement a generic  key-value API using the `ActionCache` and I'm unable to get a key without storing it into an `ObjectStore`. I should not need to store data for the only reason of passing a key to `ActionCache`.

> Insert ArrayRef into an ObjectStore to get an CASID before insert into ActionCache. You can use a InMemoryCAS, or even implement a nullCAS, which just produce the CASID and threw away the storage.

I can do that for now.

Reply:
> Does this change whether there is a concern or not?

No. Let me rephrase that. The thing I am against is adding the interface to CacheKey class that introduces the wrong behavior, which is the raw bytes stored in `CacheKey` can mean different things and can go through different amount of hash functions. `CacheKey::CacheKey(ArrayRef<uint8_t> Data)` is fine, but constructing using BLAKE3 is not.

I think a new CASContext for your wrapper ActionCache is the fitting concept to use. But I forgot that we mentioned that we should maybe put two CASContext into ActionCache to represent different contexts of keys and values. Let me know if that is too hard so we can adjust the ActionCache correspondingly.

Reply:
I mainly want the `CacheKey(ArrayRef<uint8_t> Data)` constructor to be available for clients. If that requires a `CASContext` it should be an implementation detail of the `ActionCache`.

For the purposes of this PR I'm going to be using an in-memory `ObjectStore` in order to get a key.

Reply:
See new commit: https://github.com/apple/llvm-project/pull/5355/commits/03fa5cde0aeaacda35703f0e002a85bcae00434b

Reply:
> I mainly want the CacheKey(ArrayRef<uint8_t> Data) constructor to be available for clients.

I would also be happy with `ActionCache` methods that I can use to form keys, it's not strictly necessary that the `CacheKey` must be formed using context-free constructors.